File: ruby/muse/algorithms/sorting/QuickSort.rb
module QuickSort

    def self.partition(arr, lo, hi)
        if lo == hi
            return
        end

        pivot = arr[lo]

        left = lo
        right = hi - 1

        until left == right
            right.downto(left + 1) do |i|
                if arr[i] < pivot
                    arr[left] = arr[i]
                    arr[i] = pivot
                    break
                end
                right -= 1
            end
            left.upto(right - 1) do |i|
                if arr[i] > pivot
                    arr[right] = arr[i]
                    arr[i] = pivot
                    break
                end
                left += 1
            end
        end

        partition(arr, lo, left)
        partition(arr, left + 1, hi)
    end

    def self.sort(arr)
        partition(arr, 0, arr.length)
    end
end

File: python/muse/algorithms/sorting/quick_sort.py
def partition(arr: list[int], lo: int, hi: int) -> int:
    pivot = arr[lo]

    i = lo
    j = hi - 1

    while i != j:
        while j > i:
            if arr[j] < pivot:
                arr[i], arr[j] = arr[j], pivot
                break
            j -= 1
        while i < j:
            if arr[i] > pivot:
                arr[j], arr[i] = arr[i], pivot
                break
            i += 1

    return i


def sort(arr: list[int]) -> None:
    sort_with_range(arr, 0, len(arr))


def sort_with_range(arr: list[int], lo: int, hi: int) -> None:
    if lo >= hi:
        return

    p = partition(arr, lo, hi)

    sort_with_range(arr, lo, p)
    sort_with_range(arr, p + 1, hi)
